### PR TITLE
[gui2] Language Server binary protocol, and loading of visualization data

### LINKED ISSUE #7995: Make engine connections more robust against occasional timeouts
In current state, the language server connection tends to often return timeouts, including during the initialization. We need to be able to handle those scenarios more gracefully with request retries.

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Depends on #7773.

- Implements binary WebSocket protocol (data protocol)
- Performs some editor initialization (the bare minimum so that visualizations work)
- Adds event handlers to receive visualization data updates

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

**Comment by somebody1234** [comment]
@Frizi I *think* this is more or less ready for review too.

**Comment by somebody1234** [comment]
:sob: the viz shows below other nodes

File: app/gui2/src/components/GraphNode.vue
Comment (Frizi) [inline app/gui2/src/components/GraphNode.vue]: Detaching is important. Otherwise we will be constantly adding work on the language server side. You can us e`onCleanup` of `watchEffect` here to make it a bit simpler. Just make sure to call it before first `await`.

File: app/gui2/shared/websocket.ts
Comment (Frizi) [inline app/gui2/shared/websocket.ts:1]: Why is this file necessary? Can't we use `lib0/websocket` directly instead?

File: app/gui2/src/components/GraphEditor.vue
Comment (Frizi) [inline app/gui2/src/components/GraphEditor.vue]: Inconsistent prop casing. Also I'm not a fan of passing `dataServer` and related things to the nodes here. The low level part of managing the execution context and visualization messages should be moved out of the components.

File: app/gui2/src/components/GraphNode.vue
Comment (Frizi) [inline app/gui2/src/components/GraphNode.vue]: As before, I don't like that this low level communication is done directly in the vue component's code. I think here we should at most have something like:
```ts
const visualizationData = useVisualizationData(visualizationConfigRef)
```

File: app/gui2/shared/binaryProtocol.ts
Comment (Frizi) [inline app/gui2/shared/binaryProtocol.ts]: Could you describe a bit more how it was generated?
- the tool and version used (i assume flatc, but not sure which one)
- what kind modifications were made

File: app/ide-desktop/lib/dashboard/esbuild-config.ts
Comment (Frizi) [inline app/ide-desktop/lib/dashboard/esbuild-config.ts]: That change seems unnecessary. If you have type errors, double-check if you didn't install two different esbuild versions.

File: app/ide-desktop/lib/dashboard/src/authentication/src/error.ts
Comment (Frizi) [inline app/ide-desktop/lib/dashboard/src/authentication/src/error.ts]: Again, types should match just fine here on their own. Make sure you have the dependencies installed correctly.

image
image

File: package-lock.json
Comment (Frizi) [inline package-lock.json:1]: Those lockfile changes seems very excessive. The only thing that should change is `lib0` version. Try reverting it and doing `npm install` again.

**Review by Frizi — CHANGES_REQUESTED** [review]

File: app/gui2/src/components/GraphNode.vue
Comment (Frizi) [reply app/gui2/src/components/GraphNode.vue]: Detaching is important. Otherwise we will be constantly adding work on the language server side. You could use `onCleanup` of `watchEffect` here to make it a bit simpler. Just make sure to call it before first `await`.

Though it probably should be mostly handled elsewhere anyway (see next comments).

File: app/gui2/shared/websocket.ts
Comment (somebody1234) [reply app/gui2/shared/websocket.ts:1]: it doesn't support sending binary messages, and uses a custom JSON-based ping/pong (which the server must support, and i believe our servers don't support it)

(see https://github.com/dmonad/lib0/issues/73)

File: app/gui2/src/components/GraphEditor.vue
Comment (somebody1234) [reply app/gui2/src/components/GraphEditor.vue]: i was assuming `v-bind` props were supposed to stay as `kebab-case` - can be changed 

re: the low level part of managing the execution context and visualization messages - yeah, this will probably be addressed in adam's PR, i'll probably wait for that to land before merging that into this PR.

i'll see what i can do before that gets merged in though

**Comment by somebody1234** [comment]
ok so the viz should be working (inconsistently as usual though)

... really not sure about how the execution context is currently set up. it might be nice to refactor the execution context setup in another PR?

**Comment by somebody1234** [comment]
note: this is still *not tested* with a new build of PM, so it may be broken.

**Review by Frizi — APPROVED** [review]

File: app/gui2/shared/languageServerTypes.ts
Comment (farmaazon) [inline app/gui2/shared/languageServerTypes.ts]: It _is_ defined as an optional field in the document, and there is a difference between an empty array and a missing field.

File: app/gui2/src/components/GraphEditor.vue
Comment (somebody1234) [inline app/gui2/src/components/GraphEditor.vue]: @Frizi just want to confirm that this reordering of imports is intentional?

File: app/gui2/src/components/visualizations/LoadingVisualization.vue
Comment (somebody1234) [inline app/gui2/src/components/visualizations/LoadingVisualization.vue]: incorrect name

File: app/gui2/src/components/visualizations/SQLVisualization.vue
Comment (somebody1234) [inline app/gui2/src/components/visualizations/SQLVisualization.vue:7]: are these `as const`s really needed?

File: app/gui2/src/components/GraphEditor.vue
Comment (Frizi) [reply app/gui2/src/components/GraphEditor.vue]: Yes. I found that using separate import lists keeps the IDE auto-format breaking. The only working configuration is if all imports are in the first `script` block in a file. To reduce friction, I think we should simply accept that as a rule.

File: app/gui2/src/components/GraphNode.vue
Comment (somebody1234) [inline app/gui2/src/components/GraphNode.vue]: idk whether a wrapper function that only contains a single function call and nothing else, is all that useful

File: app/gui2/src/components/visualizations/SQLVisualization.vue
Comment (Frizi) [reply app/gui2/src/components/visualizations/SQLVisualization.vue:7]: It makes the type true to what's specified in `VisualizationModule`. I attempted to remove the `as any` from `builtinVisualizationImports` list, and I sort of succeeded with it, at least in IDE. Unfortunately it still breaks during external typecheck, so I've kept the any casts for now. But at least I've made some progress towards that.

File: app/gui2/src/components/VisualizationContainer.vue
Comment (somebody1234) [inline app/gui2/src/components/VisualizationContainer.vue]: `modelValue` + `update:modelValue` may be more conventional than `value` + `update:value`

File: app/gui2/src/util/shortcuts.ts
Comment (somebody1234) [inline app/gui2/src/util/shortcuts.ts:325]: ?!

File: app/gui2/ydoc-server/edits.ts
Comment (somebody1234) [inline app/gui2/ydoc-server/edits.ts]: should `VisualizationMetadata` be a `type` import?

File: app/gui2/ydoc-server/languageServerSession.ts
Comment (somebody1234) [inline app/gui2/ydoc-server/languageServerSession.ts]: commented code

File: app/gui2/src/stores/project.ts
Comment (farmaazon) [inline app/gui2/src/stores/project.ts]: I think this should not be part of public API. You always want to have just one "call" frame on the bottom of the stack - so it should be done in `create` method.

File: app/gui2/src/util/shortcuts.ts
Comment (somebody1234) [reply app/gui2/src/util/shortcuts.ts:325]: well, i think it's not *vital* but i think it'd be nice to sort of have builtin linting - it's true that this isn't useful for builtin shortcuts, but i think it'd be nice for user-defined shortcuts. that said, `console.warn` isn't very visible, but i don't think we have a well-defined place to put status notifications right now

File: app/gui2/src/util/net.ts
Comment (farmaazon) [inline app/gui2/src/util/net.ts:47]: Shan't we print an error log here?

File: app/gui2/src/util/net.ts
Comment (farmaazon) [inline app/gui2/src/util/net.ts]: I'd be nice to see tests here.

Even for exponential backoff, here however I don't know how to handle waits.

**Review by farmaazon — APPROVED** [review]
I''d like to see some tests of

File: app/gui2/src/util/net.ts
Comment (Frizi) [reply app/gui2/src/util/net.ts:47]: This is a low-level utility. The adequate logs are already printed by its users whenever it makes sense.

File: app/gui2/src/util/net.ts
Comment (Frizi) [reply app/gui2/src/util/net.ts]: Waits should actually be pretty easily testable. It's all a `setTimeout` under the hood, and those are already [mocked by vitest](https://vitest.dev/guide/mocking.html#timers). I'll add some tests then.

File: app/gui2/shared/languageServerTypes.ts
Comment (Frizi) [reply app/gui2/shared/languageServerTypes.ts]: I don't see any explanation of a difference between missing value and an empty array in the protocol doc. Do you know what it is?

File: app/gui2/src/util/shortcuts.ts
Comment (Frizi) [reply app/gui2/src/util/shortcuts.ts:325]: This whole lint was incorrect once you used any overrides in `normalizedKeyboardSegmentLookup`. E.g. using `OsDelete` or `Space` (which I added, since it was impossible to use `' '` due to trim in decomposition logic). After all, `decomposed.key` is simply not necessarily what you actually want to write in your binding.

We could add normalization and a lint again, but it would require a bit more changes to the whole decomposition step.

Comment (somebody1234) [reply app/gui2/src/util/shortcuts.ts:325]: ah yeah, good point. i think this was written before the overrides and i kinda just forgot that that would invalidate this

**Pull request merged**
